pkg/rebalancer/rebalancer.go

```diff
@@ -0,0 +1,241 @@
+	// balancing the pods.
+	err = wait.Poll(10*time.Second, 5*time.Minute, func() (bool, error) {
+		klog.V(4).Info("Waiting until workload %s in namespace %s becomes correctly balanced.", workload.LabelSelector, workload.Namespace)
+		return r.WorkloadCorrectlyBalanced(ctx, workload)
+	})
+	return err == nil, err
+}
+
+// rebalanceWorkload rebalances the given pod on a different node. To make sure
+// that the pod will be rescheduled on a different node, this function will
+// first cordon the node in which the pod is scheduled. Then it will verify
+// whether the PVC attached to the pod needs to be deleted or not. Only pods
+// that have PVCs with a zonal topology need to be deleted since they might
+// prevent the pod to be schedule on a node that is located in a different
+// availability zone.
+func (r *Rebalancer) rebalanceWorkload(ctx context.Context, pod *v1.Pod, pvc *v1.PersistentVolumeClaim) error {
+	node, err := r.client.CoreV1().Nodes().Get(ctx, pod.Spec.NodeName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	err = drain.RunCordonOrUncordon(r.drainer, node, true)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err := drain.RunCordonOrUncordon(r.drainer, node, false)
+		if err != nil {
+			klog.Errorf("Couldn't uncordon node %v: %v.", pod.Spec.NodeName, err)
+		}
+	}()
+
+	pv, err := r.client.CoreV1().PersistentVolumes().Get(ctx, pvc.Spec.VolumeName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	// Do not delete the PVC if the storage provider hasn't set the topology.kubernetes.io/zone label on the PV.
+	// In most cases, when the PV isn't zonal, pods can access it from a node in a different AZ so we don't need to delete it.
+	if _, ok := pv.GetLabels()[zonalTopologyAnnotation]; ok {
+		klog.V(2).Infof("Deleting PersistentVolumeClaim %s/%s.", pvc.Namespace, pvc.Name)
+		err = r.client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Delete(ctx, pvc.Name, metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	klog.V(2).Infof("Deleting pod %s/%s.", pod.Namespace, pod.Name)
+	return r.client.CoreV1().Pods(pod.Namespace).Delete(ctx, pod.Name, metav1.DeleteOptions{})
+}
+
+func (r *Rebalancer) resourcesToRebalance(ctx context.Context, pods []v1.Pod) ([]v1.Pod, []v1.PersistentVolumeClaim, error) {
+	var (
+		annotatedPVCs []v1.PersistentVolumeClaim
+		podsToDelete  []v1.Pod
+	)
+	for _, pod := range pods {
+		for _, vol := range pod.Spec.Volumes {
+			if vol.PersistentVolumeClaim == nil {
+				continue
+			}
+			pvc, err := r.client.CoreV1().PersistentVolumeClaims(pod.Namespace).Get(ctx, vol.PersistentVolumeClaim.ClaimName, metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				break
+			}
+			if err != nil {
+				return nil, nil, err
+			}
+
+			dropPVC, ok := pvc.Annotations[dropPVCAnnotation]
+			if !ok {
+				break
+			}
+			if dropPVC == "yes" {
+				annotatedPVCs = append(annotatedPVCs, *pvc)
+				podsToDelete = append(podsToDelete, pod)
+			}
```

Comment:
Because we are not deleting anything here, we are only gathering the list of pods/pvcs that should be deleted, and later when it comes to their deletion we are logging.